### PR TITLE
[NETBEANS-5427] Replace usage of JDK internal, or in-house Base64 codecs by java.util.Base64

### PR DESCRIPTION
**Pull request by piercemar**
JEP 396 introduced by JDK 16 forbids access to internal classes.
JDK 1.8 provided a public alternative to internal com.sun.org.apache.xerces.internal.impl.dv.util.Base64 in the form of java.util.Base64.

### CODE REVIEW

**Comment by geertjanw** [comment]
Nicely done, thank you.

And everything works exactly as before with this change?

**Comment by piercemar** [comment]
As far as I could test, yes!
I built my branch and ran "ant tryme", activated the Tomcat plugin, added a new Tomcat 9 instance, started it, opened the "webapps" node (which is where it failed before), and stopped.
Everything went fine on JDK 11.

Unfortunately "ant tryme" segfaults on JDK16 and I can't figure out why (yet).

File: enterprise/tomcat5/src/org/netbeans/modules/tomcat5/deploy/TomcatManagerImpl.java [hunk 1]
```diff
@@ -530,7 +530,7 @@ public synchronized void run () {
                 // Set up an authorization header with our credentials
                 TomcatProperties tp = tm.getTomcatProperties();
                 String input = tp.getUsername () + ":" + tp.getPassword ();
-                String auth = new String(Base64.encode(input.getBytes()));
+                String auth = new String(Base64.getEncoder().encode(input.getBytes()));
```
Comment (matthiasblaesing) [inline enterprise/tomcat5/src/org/netbeans/modules/tomcat5/deploy/TomcatManagerImpl.java:533]: Please don't use the String constructor (especially not without a charset), but use `encodeToString`

Comment (piercemar) [reply enterprise/tomcat5/src/org/netbeans/modules/tomcat5/deploy/TomcatManagerImpl.java]: Done as suggested. Thanks

**Comment by sdedic** [comment]
Note: there are other impls and usages of Base54 encoders throughout the codebase:
- `org.netbeans.modules.proxy.Base64Encoder`
- `org.netbeans.modules.diff.builtin.Base64`
- `org.netbeans.modules.db.util.Base64`

Shouldn't they be reviewed, deprecated and/or replaced by j.u.Base64 usage ?

**Comment by matthiasblaesing** [comment]
Indeed it would be good to remove the referenced implementation or at least make them shims to call into the JDK implementation (if they are API). I had a quick look at all three implementations though and all of them are full implementations and don't call into restricted JDK apis.

@piercemar would you be willing to have a look?

Two other things:

- The author information is invalid - the email address is not valid (looks like an edit on github, if you work on github please ensure, that your full name is publically visible _and_ and valid public emailaddress is set).
- it is assumed, that you had a look at the license statement of the files you changed and agree, that your changes are donated to the ASF.

**Comment by piercemar** [comment]
@matthiasblaesing, sure I can have a look.
About the author information, I'm no git(hub) guru. Here's how I plan to fix it :
* Disable e-mail privacy in GH settings
* setup my public e-mail in my git config
* squash my commits into one
* amend author information
* force push to my branch

But it's going to affect this PR history. Is it good ?

**Comment by matthiasblaesing** [comment]
Should work, though I don't trust anything github offers regarding working with history. They try to do too much magic. Feel free to use this PR as a playground - it is simple enough, that if all fails, can be redone.

**Comment by piercemar** [comment]
I hope I got it right this time.
Base64Encoder was more broadly used than I first thought.
I replaced trivial usage directly by j.u.Base64, but for other cases, I first extracted the call to in-house codecs into a package-private static method, so i could create a unit test for it, and then replaced the implementation by j.u.Base64, ensuring there's no regression due to different encoding or other subtleties.

File: ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java [hunk 2]
```diff
@@ -1,117 +0,0 @@
...
-package org.netbeans.modules.proxy;
-
-import java.io.ByteArrayOutputStream;
-
-/**
- * Bas64 encode utility class.
- *
- * @author Maros Sandor
- */
-public class Base64Encoder {
```
Comment (matthiasblaesing) [inline ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java:29]: This file is part of the API of `versioning.util`, so it should be retained (this is a good candiate to create a shim that calls into the base64 encoder).

**Review by matthiasblaesing — COMMENTED** [review]
Thank you - I left one inline comment. The author information is still missing a valid email address (hint: run `git log -n 1` in the checkout, which will show all information for the commit)

File: ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java (hunk 2 above)
Comment (piercemar) [reply ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java:29]: Shouldn'it be depprected then ?

Comment (matthiasblaesing) [reply ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java:29]: Deprecation sounds sane to me. I'm not aware of additional attributes for it in the context of netbeans.

Comment (piercemar) [reply ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java:29]: Base64Encoder restored, shimmed and deprecated.

**Comment by piercemar** [comment]
Author information fixed (had to force push again)

File: ide/subversion/src/org/netbeans/modules/subversion/ui/diff/ExportDiffAction.java [hunk 3]
```diff
@@ -351,8 +350,29 @@ private String exportBinaryFile(File file) throws IOException {
         }
         sb.append("MIME: application/octet-stream; encoding: Base64; length: ").append(file.canRead() ? file.length() : -1); // NOI18N
         sb.append(System.getProperty("line.separator")); // NOI18N
-        sb.append(Base64Encoder.encode(baos.toByteArray(), true));
+        sb.append(encodeToWrappedBase64(baos.toByteArray()));
         sb.append(System.getProperty("line.separator")); // NOI18N
         return sb.toString();
     }
+
+    static String encodeToWrappedBase64(byte[] data) {
```
Comment (sdedic) [inline ide/subversion/src/org/netbeans/modules/subversion/ui/diff/ExportDiffAction.java:358]: Consider using MIME encoder with specified length.

**Review by sdedic — CHANGES_REQUESTED** [review]
At least the cases where the decoder gracefully ignores spaces/newlines should be fixed.

**Comment by geertjanw** [comment]
Changing milestone to 12.5.

**Comment by marcorobiati** [comment]
Hi, sorry if answer is obvious but I'm new to apache and netbeans schedules and procedures, why was this shifted to 12.5?

**Comment by geertjanw** [comment]
Just my estimation that it won't be completed in time for 12.4. If it proves otherwise, we'll include it.

File: ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java [hunk 4]
```diff
@@ -16,102 +16,54 @@
...
-import java.io.ByteArrayOutputStream;
+import java.util.Base64;
 
 /**
  * Bas64 encode utility class.
  *
  * @author Maros Sandor
+ * @deprecated prefer java.util.Base64 instead
  */
+@Deprecated
```
[the remark this answers is not in the document]
Comment (piercemar) [reply ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java:29]: Sure, do as you feel appropriate. I'm a newbie contributor, so I'll happily welcome any help on the administrative part ;)

File: ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java [hunk 5]
```diff
@@ -16,102 +16,54 @@
...
-            sb.append("=");
-        } else if (i == length - 1) {
-            int d = (((int) data[i]) & 0xFF) << 16;
-            sb.append(characters[(d >> 18) & 0x3F]);
-            sb.append(characters[(d >> 12) & 0x3F]);
-            sb.append("==");
+    public static String encode(byte[] data, boolean useNewlines) {
+        final String encoded = Base64.getEncoder().encodeToString(data);
+        if (useNewlines) {
+            return wrapText(encoded, 60, System.getProperty("line.separator"));
```
[the remark this answers is not in the document]
Comment (piercemar) [reply ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java:42]: thanks for the hint, would be cleaner indeed

File: ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java [hunk 6]
```diff
@@ -16,102 +16,54 @@
...
-                return 62;
-            case '/': 
-                return 63;
-            case '=': 
-                return 0;
-            default:
-                throw new RuntimeException("Base64: unexpected code: " + c);
-        }
+    public static byte[] decode(String s) {
+        return Base64.getDecoder().decode(s);
```
[the remark this answers is not in the document]
Comment (piercemar) [reply ide/versioning.util/src/org/netbeans/modules/proxy/Base64Encoder.java:66]: Good catch, I'll fix that

**Comment by piercemar** [comment]
@sdedic, I fixed the places where non-printable characters were accepted as per your recommendation. Thanks for catching this.

**Pull request merged**